optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/ConsecutiveIntervalInfoImpl.java

```diff
@@ -0,0 +1,214 @@
+import java.util.TreeSet;
+import java.util.function.BiFunction;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.optaplanner.examples.common.experimental.api.ConsecutiveIntervalInfo;
+import org.optaplanner.examples.common.experimental.api.IntervalBreak;
+import org.optaplanner.examples.common.experimental.api.IntervalCluster;
+
+public class ConsecutiveIntervalInfoImpl<IntervalType_, PointType_ extends Comparable<PointType_>, DifferenceType_ extends Comparable<DifferenceType_>>
+        implements
+        ConsecutiveIntervalInfo<IntervalType_, PointType_, DifferenceType_> {
+    private final NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> clusterStartSplitPointToCluster;
+    private final NavigableSet<IntervalSplitPoint<IntervalType_, PointType_>> splitPointSet;
+    private final NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>> clusterStartSplitPointToNextBreak;
+    private final Iterable<IntervalCluster<IntervalType_, PointType_, DifferenceType_>> intervalClusterIterable;
+    private final BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction;
+    private final Iterable<IntervalBreak<IntervalType_, PointType_, DifferenceType_>> breaksIterable;
+
+    public ConsecutiveIntervalInfoImpl(TreeSet<IntervalSplitPoint<IntervalType_, PointType_>> splitPointSet,
+            BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction) {
+        clusterStartSplitPointToCluster = new TreeMap<>();
+        clusterStartSplitPointToNextBreak = new TreeMap<>();
+        intervalClusterIterable = new MapValuesIterable<>(clusterStartSplitPointToCluster);
+        breaksIterable = new MapValuesIterable<>(clusterStartSplitPointToNextBreak);
+        this.splitPointSet = splitPointSet;
+        this.differenceFunction = differenceFunction;
+    }
+
+    protected void addInterval(Interval<IntervalType_, PointType_> interval) {
+        NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> intersectedIntervalClusterMap =
+                clusterStartSplitPointToCluster.subMap(
+                        ObjectUtils.defaultIfNull(clusterStartSplitPointToCluster.floorKey(interval.getStartSplitPoint()),
+                                interval.getStartSplitPoint()),
+                        true, interval.getEndSplitPoint(), true);
+
+        // Case: the interval cluster before this interval does not intersect this interval
+        if (!intersectedIntervalClusterMap.isEmpty()
+                && intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey()).getEndSplitPoint()
+                        .compareTo(interval.getStartSplitPoint()) < 0) {
+            intersectedIntervalClusterMap = intersectedIntervalClusterMap.subMap(intersectedIntervalClusterMap.firstKey(),
```

Comment:
`firstKey` is used several times in this method. For non-trivial methods, you should really consider caching the result.
(Technically this should be a trivial method, but for some reason, it really is not, as my profiling showed when we were doing the consecutive sequences.)

Reply:
Done (technically already cached as oldStart, but I wasn't using it)